### S02 T04: Pràctica amb programació numèrica

* **Exercici 1**

Crea una funció que donat un Array d’una dimensió, et faci un resum estadístic bàsic de les dades. Si detecta que l’array té més d’una dimensió, ha de mostrar un missatge d’error.

In [7]:
import numpy as np
import random 

In [8]:
# crearem una funció que rebi com a paràmetre un array i ens torni un string amb les dades estadístiques o un missatge 
# d'error si la dimensio no es 1

def array_statistics(array): 
    
    size = array.size 
    mean = np.mean(array)
    median = np.median(array)
    variance = np.var(array)
    std_deviation = np.std(array)
    dimensions = array.ndim 
    
    if dimensions == 1: 
        data = "Tamany : " + str(size) + "\n"+"Mitjana: " + str(mean) + "\n" + "Median: " + str(median) + "\n" + \
        "Variança: "+ str(variance) +"\n"+"Desviació estàndar: "+ str(std_deviation) + "\n"+"Dimensions: "+ str(dimensions)
    else:  
        data = "Error, la matriu ha de ser d'una dimensió."
    return data

In [9]:
# Generarem un array d'una dimensió amb 15 números aleatoris del 1 al 50

random_array = np.random.randint(1,51,15)
data = array_statistics(random_array)
print(data)

Tamany : 15
Mitjana: 26.533333333333335
Median: 32.0
Variança: 195.44888888888892
Desviació estàndar: 13.98030360503265
Dimensions: 1


In [10]:
# Amb la funció reshape canviarem les dimensions de la matriu per veure si ens torna un missatge d'error 

random_array = random_array.reshape(3,5)
data = array_statistics(random_array)
print(data)

Error, la matriu ha de ser d'una dimensió.


* **EXERCICI 2**


Crea una funció que et generi un quadrat NxN de nombres aleatoris entre el 0 i el 100.

In [11]:
import numpy as np 
import random

In [12]:
def array_generator(number): 
# primer generem un array amb nombres aleatoris del 0 al 100 i després la canviarem amb la funció reshape perque tingui 
# 2 dimensions 

    random_square = np.random.randint(0,101,number**2).reshape(number, number)
    return random_square

number = int(input("Escull un número sencer: "))
random_square = array_generator(number)
random_square


Escull un número sencer: 4


array([[94,  3, 73, 41],
       [13, 81, 38,  8],
       [93, 15, 19,  9],
       [89, 69, 20, 91]])

+ **EXERCICI 3**

Crea una funció que donada una taula de dues dimensions, et calculi els totals per fila i els totals per columna.

In [13]:
# farem servir la funció de numpy apply_along_axis, on 0 vol dir columnes i 1 vol dir files. 

def total_sum(array): 
    sum_of_columns = np.apply_along_axis(sum, 0, array)
    sum_of_rows = np.apply_along_axis(sum, 1, array)
    
    return "La suma de cada fila és " +str(sum_of_rows)+ " i la suma de cada columna és " +str(sum_of_columns)

array = np.random.randint(0,26, 20).reshape(5,4)
array 

array([[18, 15, 17, 11],
       [ 0, 14, 18,  1],
       [25,  9, 21, 21],
       [20, 21, 20, 17],
       [20,  9, 25,  6]])

In [14]:
totals = total_sum(array)
totals

'La suma de cada fila és [61 33 76 78 60] i la suma de cada columna és [ 83  68 101  56]'

* **EXERCICI 4**

Implementa manualment una funció que calculi el coeficient de correlació. Informa’t-en sobre els seus usos i interpretació.

El quoeficient de correlació es fa servir per veure la relació entre dos variables quantitatives. Ens serveix per respondre preguntes del tipus: "al canviar el valor d´una de les variables, canvien els valors també en l'alta?".  Aquesta mesura ens **dona informació tant de la intensitat com de la direcció de la relació**. Comprén valors entre 1 i -1: 

* Un valor menor que 0: vol dir que les variables estan associades en sentit invers. Com més pròxim a -1, major és la força d'aquesta relació invertida. Si és -1 vol dir que les variables tenen una correlació negativa perfecta. 

* Un valor major que 0 vol dir que existeix una correlació positiva. En aquest cas les variables estam associades de forma directa, com més a prop de 1 més directa és aquesta relació. Un valor de  indicaria una relació positiva perfecta. 

* Una correlació de  o pròxima a  voldria dir que no existeix relació entre les dos variables. 

Ara calcularem la fòrmula i veurem alguns exemples pràctics. 

Entenc que sen's demana fer de forma manual tots els càlculs possibles, per lo tant faré servir la fòrmula: 

<div>
<img src="formula.png" width="500"/>
</div>

Primer crearem una funció per calcular la mitjana d'un array donat, així la podrem cridar quan la necessitem: 

In [15]:
import numpy as np

In [16]:
def mean_calculation(array): 
    x_sum = 0
    size = 0
    for x in array: 
        x_sum = x_sum + x 
        size = size +1
    mean = x_sum / size
    return mean 

Ara crearem una funció que calculi les diferències de cada element de l'array respecte a la mitjana. Aquesta funció retornarà un array amb les diferències.

In [25]:
def list_of_differences(array, mean): 
    total_differences = array -mean 
    return total_differences  


Ara hem de calcular la suma dels productes entre dos arrays donades i tindrem el numerador de la formula (la suma dels productes de les seves diferències). 

In [28]:
def sum_of_products(array_x, array_y): 
    mean_x = mean_calculation(array_x)
    mean_y = mean_calculation(array_y)
    total_differences_x = list_of_differences(array_x , mean_x)
    total_differences_y = list_of_differences(array_y, mean_y)  
    
    total_list = total_differences_x*total_differences_y
    
    total_sum = 0
    for i in total_list: 
        total_sum = total_sum + i 
    return total_sum

Ara que ja tenim el numerador de la fòrrmula, haurem de calcular el denominador. Primer calcularem la suma de les diferències al quadrat: 


In [49]:
def square_differences(array): 
    mean = mean_calculation(array)
    # ja tenim la funció que retorna una array de diferències, la cridem i la recorrerem amb un bucle for on calcularem
    # el quadrat de cada element i l'anirem sumant. La funció retornarà la suma de tots els elements al quadrat.
    differences = list_of_differences(array, mean)
    squares = 0
    for i in differences: 
        squares = squares + i**2
    return squares

Ara ja podem definir la funció per calcular la correlació: 

In [30]:
def correlation_quoefficient(array_x, array_y): 
    import math 
    
    # cridem la funció que conté el nominador de la fòrmula: 
    nominator = sum_of_products(array_x, array_y)
    
    # Fem els càlculs per definir el denominador fent servir la funció que retorna la suma de diferències al quadrat: 
    squares_x = square_differences(array_x)
    squares_y = square_differences(array_y)
    product_of_squares = squares_x*squares_y
    root_of_squares = math.sqrt(product_of_squares)
    
    # ara ja tenim tots els calculs per fer la formula: 
    correlation = nominator/root_of_squares
    
    return correlation

Per probar la fòrmula, primer calcularem la correlació entre dos arrays aleatoris: 

In [53]:
array_x = np.random.randint(1,51,10)
array_y = np.random.randint(1,51,10)
correlation = correlation_quoefficient(array_x, array_y)
print(correlation)
# imprimiré també la funció que ens dona numpy per comparar que el resultat sigui el mateix i que la fòrmula està be: 
print(np.corrcoef(array_x, array_y))

0.08876374785979799
[[1.         0.08876375]
 [0.08876375 1.        ]]


Veiem que la correlació es molt baixa, 0.089 (8,9 %), aixó té sentit ja que estem buscant la relació entre dos arrays fets amb nombres aleatoris, té sentit que aquests no tinguin cap correlació. De tota manera si probem el resultat vàries vegades la correlació pot sortir més alta, per lo tant, entenem que la correlació no hauria de ser l'únic indicador que tindrem en compte a l´hora de definir la relació entre dos variables. 

Crearem un array del 1 al 5 i un altre amb els valors que siguin el doble del primer array, a veure quin és el resultat: 

In [45]:
array_x = np.array((1,2,3,4,5))
array_y = np.array((2,4,6,8,10))
correlation = correlation_quoefficient(array_x, array_y)
correlation

1.0

La correlació es perfecta i positiva, aixó es perque els elements de l'array y son exactament el doble dels elements de x, per lo tant, la correlació en aquest cas és 100% positiva. 

Ara farem un exemple de correlació 100% negativa: 

In [46]:
array_x = np.array((1,2,3,4,5))
array_y = np.array((10,8,6,4,2))
correlation = correlation_quoefficient(array_x, array_y)
correlation

-1.0

També té sentit, perque els elements de l'array y decreixen al mateix rati que augmenten els del array x, per lo tant el coeficient de correlaciò entre els dos arrays és 100% negativa. 

Si cambiem alguns elements dels arrays anteriors de forma manual: 

In [48]:
array_x = np.array((1,5,3,4,7))
array_y = np.array((3,4,6,8,10))
correlation = correlation_quoefficient(array_x, array_y)
correlation

0.7418253689708787

Veiem que el coeficient de correlació ha baixat, ara és del 74%, els elements dels arrays ja no tenen una correlació perfecta. 